File: skript-parser-java-impl/src/main/java/io/github/syst3ms/skriptparser/registration/ExpressionInfo.java
package io.github.syst3ms.skriptparser.registration;

import io.github.syst3ms.skriptparser.pattern.PatternElement;
import io.github.syst3ms.skriptparser.types.PatternType;
import io.github.syst3ms.skriptparser.types.Type;
import org.jetbrains.annotations.NotNull;

import java.util.List;

/**
 * Represents something that returns something and has PATTERNS which
 * may result it being matched
 */
public class ExpressionInfo<C, T> extends SyntaxInfo<C> {
    private PatternType<T> returnType;

    public ExpressionInfo(Class<C> c, List<PatternElement> patterns, Type<T> returnType, boolean isSingle, int priority) {
        super(c, patterns, priority);
        this.returnType = new PatternType<>(returnType, isSingle);
    }

    public PatternType<T> getReturnType() {
        return returnType;
    }
}

File: skript-parser-kotlin/src/main/kotlin/io/github/syst3ms/skriptparser/registration/ExpressionInfo.kt
package io.github.syst3ms.skriptparser.registration

import io.github.syst3ms.skriptparser.classes.Expression
import io.github.syst3ms.skriptparser.pattern.PatternElement

/**
 * Represents something that returns something and has patterns which
 * may result it being matched
 */
class ExpressionInfo<C : Expression<out T>, out T>(c: Class<C>, patterns: List<PatternElement>, val returnType: Type<out T>) : SyntaxInfo<C>(c, patterns)